add technote on interfaces (#265)
apollographql/docs: src/content/technotes/TN0018-abstract-types.mdx

---
title: GraphQL abstract types
id: TN0018
tags: [schema design, federation]
---

> ## Interfaces are for clients

Reviewer: ```suggestion
```

> Backend developers using languages like Java may be tempted to overuse interfaces to enforce a similar shape across distinct types.

Reviewer: ````suggestion
GraphQL interfaces enable schema fields to return one of multiple object types, all of which must _implement_ that interface. To implement an interface, an object type must define all fields that are included in that interface (otherwise, the schema is invalid):

```graphql
interface Media {
  title: String!
}

type Book implements Media {
  title: String!
  author: String!
}
```

Because interfaces can enforce this requirement on implementing object types, it can be tempting to create interfaces _purely_ to enforce field definitions. The following schema requires a `name` field on all of its object types by declaring that those types implement the `Nameable` interface:
````

> In the following schema, we're ensuring that all types can have a name by declaring that they implement the same interface.

Reviewer: ```suggestion
```

> ```graphql title="❌ Don't do this"

Reviewer: ````suggestion

<p style="margin-bottom: 0;">❌</p>

```graphql title="schema.graphql"
````

Reviewer: (Nothing wrong with how this was, this is just how the docs do this sorta thing elsewhere for consistency's sake 🤷 )

interface Nameable {
  name: String
}

type Cat implements Nameable {
  name: String
  meowVolume: Int
}

type Dog implements Nameable {
  name: String
  barkVolume: Int
}

type Owner implements Nameable {
  name: String
  cats: [Cat]
  dogs: [Dog]
}

type Query {
  owners: [Owner]
}
```

> But note that at no point is `Nameable` used as a return type for a field. There's no client use case for fetching a polymorphic list of Cats, Dogs, and Owners.

Reviewer: ```suggestion
**This is not a recommended use for interfaces!** Note that in this schema, `Nameable` is never used as the return type for a field. This means that clients have no way to take advantage of this polymorphic relationship in their operations.
```

> The lack of valid client use cases indicates a code smell. The `Nameable` interface is unnecessary and it may cause problems when making changes to your schema in the future. And as you'll see later in this document, spreading interfaces across subgraphs has its own challenges, so removing unnecessary interfaces will often make life simpler.

Reviewer: ```suggestion
The lack of valid client use cases indicates a code smell. The `Nameable` interface is unnecessary, and it might cause problems when making changes to your schema in the future. And because distributing an interface across subgraphs presents its own challenges, removing unnecessary interfaces usually makes life simpler.
```

> In the next example, we've specified that both the `Cat` and `Dog` types implement the `Pet` interface so that we can return a polymorphic list of both Cats and Dogs in the `Owner.pets` field. Now that there's a valid client use case for polymorphism, using an interface is justified.

Reviewer: ```suggestion
In the next example, we specify that both the `Cat` and `Dog` types implement the `Pet` interface so that we can return a polymorphic list of both Cats and Dogs in the `Owner.pets` field. Now that there's a valid client use case for polymorphism, using an interface is justified.
```

> ```graphql title="✅ A better use of interfaces"

Reviewer: ````suggestion
<p style="margin-bottom: 0;">✅</p>

```graphql title="schema.graphql"
````

interface Pet {
  name: String
}

type Cat implements Pet {
  name: String
  meowVolume: Int
}

type Dog implements Pet {
  name: String
  barkVolume: Int
}

type Owner {
  name: String
  pets: [Pet]
}

type Query {
  owners: [Owner]
}
```

> ## Consider implications of adding new types

Reviewer: ```suggestion
## Implications of adding new implementing types
```

> When a client selects an abstract type in an operation, they usually enumerate the concrete types they're interested in with "conditional fragments".

Reviewer: ```suggestion
When a client includes a field with an abstract return type in an operation, they usually enumerate which fields from _concrete_ types they're interested in with "conditional fragments," like so:
```

```graphql
query OwnersAndPets {
  owners {
    name
    pets {
      name
      ... on Cat {
        meowVolume
      }
      ... on Dog {
        barkVolume
      }
    }
  }
}
```

> > If `Pet` was a union instead of an interface (`union Pet = Cat | Dog`), the client _must_ select the concrete types. It's not possible to select only the `name` field even though the types have that field in common.

Reviewer: ```suggestion
> If `Pet` were a _union_ instead of an interface (`union Pet = Cat | Dog`), then _all_ requested fields would need to be included in conditional fragments. It wouldn't be possible to include the `name` field outside of a fragment, even though the two types both happen to define that field.
```

> If your API adds additional possible types to the abstract type, client must change their operations to select data from those new types. While not a _breaking_ change, this situation warrants some defensive programming.

Reviewer: ```suggestion
If your API adds additional possible types to the abstract type, clients must change their operations to select data from those new types. While this isn't a _breaking_ change, this situation warrants some defensive programming.
```
